// js-to-ts-converter/dist/converter/add-class-property-declarations/js-class.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsClass = void 0;
const set_utils_1 = require("../../util/set-utils");
/**
 * Represents a JavaScript class found in a source file.
 */
class JsClass {
    constructor(cfg) {
        this.name = cfg.name;
        this.path = cfg.path;
        this.superclassName = cfg.superclassName;
        this.superclassPath = cfg.superclassPath;
        this.methods = cfg.methods || new Set();
        this.properties = cfg.properties || new Set();
        this.members = set_utils_1.union(this.methods, this.properties);
    }
    /**
     * String identifier for the JsClass which is a combination of its file path
     * and class name. Used to store JsClass nodes on a graphlib Graph.
     */
    get id() {
        return `${this.path}_${this.name}`;
    }
    /**
     * Retrieves the ID of the superclass JsClass instance, if the JsClass has
     * one. If not, returns undefined.
     *
     * Also returns `undefined` if the class is found to be in the node_modules
     * folder, as we don't want to attempt to parse ES5 modules.
     */
    get superclassId() {
        if (this.isSuperclassInNodeModules()) {
            // If the superclass is in the node_modules folder, we'll
            // essentially treat this JsClass as if it didn't have a superclass.
            // See `isSuperclassInNodeModules()` jsdoc for details.
            return undefined;
        }
        else {
            return this.superclassName && `${this.superclassPath}_${this.superclassName}`;
        }
    }
    /**
     * Determines if the JsClass's superclass was found in the node_modules
     * directory (i.e. it extends from another package).
     *
     * If so, we're not going to try to understand a possibly ES5 module for
     * its properties, so we'll just stop processing at that point.
     */
    isSuperclassInNodeModules() {
        return this.superclassPath === undefined;
    }
}
exports.JsClass = JsClass;
